python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py

```diff
@@ -428,19 +443,34 @@ def evaluate_asset(
                 dynamic_partitions_store=self.instance_queryer,
             ),
             to_materialize,
+            to_skip,
             to_discard,
         )
 
+    @functools.cached_property
+    def unhandled_asset_graph_subset(self) -> AssetGraphSubset:
```

Comment:
Ah yeah this should definitely be called "skipped_asset_graph_subset". The purpose is to keep track of the set of assets that are currently in the "skipped" state. This means that there was a reason to materialize them, but they were instead skipped, and therefore more work needs to be done to evaluate if those reasons for skipping are still valid.

In contrast, the "handled_root_asset_keys" tracks the inverse (the set of things which were materialized or discarded). Even though they're essentially inverses of each other, one can't be derived from the other as the total set of partitions can change over time (in addition to there being rules other than `materialize_on_missing`).

It'd be conceptually simpler to just maintain a separate asset graph subset for each rule, basically representing the set of asset partitions for which that rule was true for (on the last tick). Then each rule (materialize or skip) has the option of using that information to help calculate what's still true. In general, materialize rules will just forward along anything that is newly true + whatever was true last tick, and skip rules can be smarter about which of the things that were true last tick even need to be reevaluated (i.e. if your parent was missing last tick, and none of your parents have been updated since last tick, then that parent is still missing).

However, the issue with maintaining separate asset graphs per rule is mostly around the fact that rules are configurable, so it's somewhat tricky to just say that the cursor will maintain a map from rule to asset graph subset (what's the actual key there -- class name?).

Although, it's also a fair argument that this implementation just sort of implicitly uses that class name as a key (it's how we pick out which part of the previous AssetEvaluationRecord maps to our specific rule). The benefit is just that we have the option to change that logic without changing any serialized information in the cursor.

But in short, we're essentially emulating the experience of having a separate AssetGraphSubset per rule by doing the following:
1. Keep track of a single asset graph subset containing any asset partitions which are in the "skipped" state
2. If you aren't in the "skipped" state, then we don't care about your previous value because (as you must have gotten materialized / discarded, or had no prior value at all). If you are in the skipped state, then we can go and look at the previous evaluation record for you to see exactly what specific rules were true last tick.

Reply:
> What exactly does it mean for an asset to be in a "skipped" state? Is it the same as saying that the asset was skipped on the most recent tick? If not, when would it be different?

There's no difference in the future state of the world -- in the current state of the world, an asset can be skipped on one tick, then ignored completely on the next as we don't re-calculate the fact that it is still skipped (but the point of this PR is to remove this property, so your definition is totally fair).